### PR TITLE
Issue 372 - move tslint to eslint

### LINKED ISSUE #372: Move tslint rules to eslint
tslint is going to be deprecated in favor of eslint.

Currently, eslint is used for js tests, and tslint for typescript code.

Using tslint for both will reduce dependencies and simplify the testing workflow. This issue should cover the following:

* Migrate tslint rules to eslint (https://github.com/typescript-eslint/typescript-eslint)
* Change code if needed to comply (changes should be minimal)
* Remove any dependency of tslint

### PR DESCRIPTION
**Pull request by luisgregson**
Resolves #372 

Changes:

* add `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to enable linting of ts files
* migrate `tslint:recommended` rules to `plugin:@typescript-eslint/recommended`
* fix some issues for rules that don't work exactly as they did in tslint or where rules were previously ignored with `@ts-ignore`

Notes:
1. There are quite a few useful rules which i've disabled for consistency with the old tslint rules. I think it's worth analysing the disabled rules but i think this should be done in a separate issue once this is merged.
1. There's now 2 eslint config files (1 for js files and one for ts files). It might be useful to merge this to bring consistency of the codebase, but doing so in this PR would mean fixing a lot of linting issues, which again i think should be looked at as a separate issue.

### CODE REVIEW

File: .eslintrc-js.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{
```
Comment (angrykoala) [inline .eslintrc-js.js:1]: What is the reasoning behind this change?

File: injection_scripts/wendigo_utils.ts [hunk 2]
```diff
@@ -50,19 +50,17 @@ export default function WendigoUtilsLoader(): void {
 
                 // Based on https://github.com/capaj/proxy-date
                 (window as any).Date = new Proxy(_origDate, {
-                    construct(target: typeof Date, args: Array<any>): Date {
```
Comment (angrykoala) [inline injection_scripts/wendigo_utils.ts:53]: Lovely :+1:

File: package.json [hunk 3]
```diff
@@ -51,13 +51,17 @@
     "@types/mocha": "^8.0.3",
     "@types/node": "^14.14.2",
     "@types/puppeteer": "~3.0.2",
+    "@typescript-eslint/eslint-plugin": "^4.5.0",
+    "@typescript-eslint/parser": "^4.5.0",
     "basic-auth": "^2.0.1",
     "eslint": "^7.12.0",
+    "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-jsdoc": "^30.7.3",
```
Comment (angrykoala) [inline package.json:59]: this plugin seems to be a default plugin added by tslint-to-eslint.

We are not using jsdoc anywhere on wendigo (and don't plan to for the moment) Could you remove this dependency along with the related rules in the eslint-file, please?

File: package.json [hunk 4]
```diff
@@ -51,13 +51,17 @@
     "@types/mocha": "^8.0.3",
     "@types/node": "^14.14.2",
     "@types/puppeteer": "~3.0.2",
+    "@typescript-eslint/eslint-plugin": "^4.5.0",
+    "@typescript-eslint/parser": "^4.5.0",
     "basic-auth": "^2.0.1",
     "eslint": "^7.12.0",
+    "eslint-plugin-import": "^2.22.1",
```
Comment (angrykoala) [inline package.json:58]: I'm not completely sure what this plugin is for (in this project) as no options from the plugin docs are used in the eslint.

**Review by angrykoala — CHANGES_REQUESTED** [review]
@luisgregson The new eslint looks nice. The minor changes done to the code to comply are also a welcome improvement.

I have a couple of questions and a minor change that I would like to have it done before merge.

Great job! thanks

**Comment by angrykoala** [comment]
Also, please make this PR to the `dev` branch instead of `master` to keep the master branch in sync with latest release :)
Thanks

File: .eslintrc-js.js (hunk 1 above)
Comment (luisgregson) [reply .eslintrc-js.js:1]: I tend to prefer them because they allow comments, imports an environment variables. i wanted it to keep the `.eslintrc-ts.js` as `.js` so that the comments could be left in, and i figured i'd move this to `.js` for consistency. Happy to change it if you prefer `json` though

File: package.json (hunk 4 above)
Comment (luisgregson) [reply package.json:58]: good point. it was automatically added by `tslint-to-eslint-config` so i'd assumed the project must need it, but i've removed the package and linting still passes 👍

File: package.json (hunk 3 above)
Comment (luisgregson) [reply package.json:59]: ah yeh, good point. i've removed 👍

**Review by luisgregson — COMMENTED** [review]
I've addressed the PR comments locally. Will push up and repoint PR to dev branch

File: .eslintrc-js.js (hunk 1 above)
Comment (angrykoala) [reply .eslintrc-js.js:1]: Nah, your point is fair enough, happy to keep it as js

File: package.json (hunk 4 above)
Comment (angrykoala) [reply package.json]: Cool, thanks for checking

File: tsconfig.json [hunk 5]
```diff
@@ -18,8 +18,11 @@
         "resolveJsonModule": true
     },
     "compileOnSave": true,
+    "exclude": ["node_modules", "dist"],
     "includes": [
         "main.ts",
-        "globals.d.ts"
+        "globals.d.ts",
+        "lib",
+        "tests"
```
Comment (angrykoala) [inline tsconfig.json:26]: I think tests should not be included in the compiled project

I'm not sure why this change, I guess something to do with eslint detecting what to lint?

**Comment by angrykoala** [comment]
All the changes look good, just one question to address about a change in tsconfig

File: tsconfig.json (hunk 5 above)
Comment (luisgregson) [reply tsconfig.json]: ah yeh. I was having trouble with getting eslint to lint files from anywhere except the root dir, and i added this while debugging. It's not required so i've put it back to how it was.

**Comment by angrykoala** [comment]
Thanks, for your changes, that was pending for a while now :+1:
